src/main/kotlin/org/rust/ide/annotator/RustLiteralAnnotator.kt

```diff
@@ -0,0 +1,68 @@
+package org.rust.ide.annotator
+
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.Annotator
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.RustLiteral
+import org.rust.lang.core.psi.RustTokenElementTypes.*
+import org.rust.lang.core.psi.visitors.RustVisitorEx
+
+class RustLiteralAnnotator : Annotator {
+    override fun annotate(element: PsiElement, holder: AnnotationHolder) = element.accept(object : RustVisitorEx() {
+        override fun visitLiteral(literal: RustLiteral) {
+            // Check suffix
+            val suffix = literal.suffix
+            val possibleSuffixes = literal.possibleSuffixes
+            if (!suffix.isNullOrEmpty() && suffix !in possibleSuffixes) {
+                holder.createErrorAnnotation(literal as PsiElement, if (possibleSuffixes.isNotEmpty()) {
+                    val possibleSuffixesStr = possibleSuffixes.map { "'$it'" }.joinToString()
+                    "invalid suffix '$suffix' for ${literal.displayName}; the suffix must be one of: $possibleSuffixesStr"
+                } else {
+                    "${literal.displayName} with a suffix is invalid"
+                })
+            }
+        }
+
+        override fun visitNumericLiteral(literal: RustLiteral.Number) {
+            // TODO: Check numeric literals boundaries
+            super.visitNumericLiteral(literal)
+        }
+
+        override fun visitTextLiteral(literal: RustLiteral.Text) {
+            // Check char literal length
+            when (literal.tokenType) {
+                BYTE_LITERAL, CHAR_LITERAL -> when {
+                    literal.value.isNullOrEmpty() ->
+                        holder.createErrorAnnotation(literal as PsiElement, "empty ${literal.displayName}")
+                    literal.value!!.length > 1    ->
```

Comment:
I would prefer to bind `literal.value` to a variable here and explicitly check for `null` so as to avoid `!!`. It is obvious that `!!` is valid here, but after several rounds of refactoring it may become more obscure.